src/Microsoft.OData.Core/HttpUtils.cs

```diff
@@ -453,37 +455,82 @@ internal static string ReadTokenOrQuotedStringValue(string headerName, string he
                         break;
                     }
 
+                    if (parameterValue == null)
+                    {
+                        parameterValue = new StringBuilder();
+                        int i = start;
+                        while (i < textIndex - 1)
+                        {
+                            parameterValue.Append(headerText[i]); // initial the string builder for the unescaped string
+                            i++;
+                        }
+                    }
+
                     if (textIndex >= headerText.Length)
                     {
                         throw createException(Strings.HttpUtils_EscapeCharAtEnd(headerName, headerText, textIndex, currentChar));
                     }
 
-                    currentChar = headerText[textIndex];
+                    currentChar = headerText[textIndex]; // only save the char after '\'? not unescape it? or it's never used?
                 }
                 else
                 {
-                    if (!isQuotedString && !IsHttpToken(currentChar))
-                    {
-                        // If the given character is special, we stop processing.
-                        break;
-                    }
-
-                    if (isQuotedString && !IsValidInQuotedHeaderValue(currentChar))
+                    if (!IsValidInQuotedHeaderValue(currentChar))
                     {
                         throw createException(Strings.HttpUtils_InvalidCharacterInQuotedParameterValue(headerName, headerText, textIndex, currentChar));
                     }
                 }
 
-                parameterValue.Append(currentChar);
+                if (parameterValue != null)
+                {
+                    parameterValue.Append(currentChar);
+                }
+
                 textIndex++;
             }
 
-            if (isQuotedString && currentChar != '\"')
+            if (currentChar != '\"')
             {
                 throw createException(Strings.HttpUtils_ClosingQuoteNotFound(headerName, headerText, textIndex));
             }
 
-            return parameterValue.ToString();
+            if (parameterValue != null)
+            {
+                return parameterValue.ToString().AsMemory();
+            }
+            else
+            {
+                return headerText.AsMemory(start, textIndex - start - 1);
+            }
+        }
+
+        /// <summary>
+        /// Skips whitespace in the specified text by advancing an index to
+        /// the next non-whitespace character.
+        /// </summary>
+        /// <param name="text">Text to scan.</param>
+        /// <param name="textIndex">Index to begin scanning from.</param>
+        /// <returns>true if the end of the string was reached, false otherwise.</returns>

```

Comment:
This `<returns>` description differs from the actual return value from the method.

Reply:
#Resolved.